cli/command/trust/signer_add.go

```diff
@@ -0,0 +1,115 @@
+package trust
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/docker/cli/cli"
+	"github.com/docker/cli/cli/command"
+	"github.com/docker/cli/cli/trust"
+	"github.com/docker/cli/opts"
+	"github.com/docker/notary/client"
+	"github.com/docker/notary/tuf/data"
+	tufutils "github.com/docker/notary/tuf/utils"
+	"github.com/spf13/cobra"
+)
+
+type signerAddOptions struct {
+	keys   opts.ListOpts
+	signer string
+	images []string
+}
+
+func newSignerAddCommand(dockerCli command.Cli) *cobra.Command {
+	var options signerAddOptions
+	cmd := &cobra.Command{
+		Use:   "signer-add [OPTIONS] NAME IMAGE [IMAGE...] ",
+		Short: "Add a signer",
+		Args:  cli.RequiresMinArgs(2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			options.signer = args[0]
+			options.images = args[1:]
+			return addSigner(dockerCli, &options)
+		},
+	}
+	flags := cmd.Flags()
+	options.keys = opts.NewListOpts(nil)
+	flags.VarP(&options.keys, "key", "k", "Path to the signer key to add")
+	return cmd
+}
+
+func addSigner(cli command.Cli, options *signerAddOptions) error {
+	signerName := options.signer
+
+	// TODO@ashfall: instead of failing if no key is provided, ask user if
+	// they want to generate a key
+	if options.keys.Len() < 1 {
+		return fmt.Errorf("path to a valid key must be provided")
+	}
+
+	for _, imageName := range options.images {
+		if err := addSignerToImage(cli, signerName, imageName, options.keys.GetAll()); err != nil {
+			fmt.Fprintf(cli.Out(), "Failed to add signer to %s: %s\n", imageName, err)
+			continue
+		}
+		fmt.Fprintf(cli.Out(), "Successfully added signer: %s to %s\n", signerName, imageName)
+	}
+	return nil
+}
+
+func addSignerToImage(cli command.Cli, signerName string, imageName string, keyPaths []string) error {
+	_, ref, repoInfo, authConfig, err := getImageReferencesAndAuth(cli, imageName)
+	if err != nil {
+		return err
+	}
+
+	notaryRepo, err := trust.GetNotaryRepository(cli, repoInfo, *authConfig, "push", "pull")
+	if err != nil {
+		return trust.NotaryError(ref.Name(), err)
+	}
+
+	if err = notaryRepo.Update(false); err != nil {
+		switch err.(type) {
+		case client.ErrRepoNotInitialized, client.ErrRepositoryNotExist:
+			fmt.Fprintf(cli.Out(), "Initializing %s...\n", imageName)
+			if err := getOrGenerateRootKeyAndInitRepo(notaryRepo); err != nil {
+				return trust.NotaryError(ref.Name(), err)
+			}
+			fmt.Fprintf(cli.Out(), "Successfully initialized %q\n", notaryRepo.GetGUN().String())
```

Comment:
let's use image name instead of GUN for consistency.  The similar block in `sign.go` might need to be updated as well, can be a separate PR